coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/IntermediateTrinoSqlNodeRewriter.java

```diff
@@ -0,0 +1,104 @@
+import org.apache.calcite.sql.SqlTypeNameSpec;
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.calcite.sql.util.SqlShuttle;
+import org.apache.calcite.sql.validate.SqlValidatorUtil;
+
+import static org.apache.calcite.rel.rel2sql.SqlImplementor.*;
+import static org.apache.calcite.sql.parser.SqlParserPos.*;
+
+
+/**
+ * IntermediateTrinoSqlNodeRewriter rewrites the TrinoSqlNode representation. It transforms the sqlCalls
+ * in the TrinoSqlNode to be compatible with Trino engine in a backward compatible manner.
+ *
+ * This is achieved by visiting the TrinoSqlNode AST in a pre-order traversal manner and
+ * transforming each SqlNode (SqlCall), wherever required.
+ * The transformation may involve change in operator, reordering the operands
+ * or even re-constructing the SqlCall.
+ *
+ * NOTE: This is a temporary class which hosts certain transformations which were previously done in RelToTrinoConverter.
+ * This class will be refactored once standardized CoralIR is integrated in the CoralRelNode to trino SQL translation path.
+ */
+public class IntermediateTrinoSqlNodeRewriter extends SqlShuttle {
+  public IntermediateTrinoSqlNodeRewriter() {
+  }
+
+  @Override
+  public SqlNode visit(final SqlCall call) {
+    SqlCall transformedSqlCall = getTransformedSqlCall(call);
+    return super.visit(transformedSqlCall);
+  }
+
+  public static SqlCall getTransformedSqlCall(SqlCall sqlCall) {
+    switch (sqlCall.getOperator().kind) {
+      case EQUALS:
+        return getTransformedEqualsOperatorSqlCall(sqlCall);
+      case SELECT:
+        return getTransformedSqlSelectSqlCall(sqlCall);
+      default:
+        return sqlCall;
+    }
+  }
+
+  // Append TryCast operator to both operands to cast each operand's data type to VARCHAR
+  private static SqlCall getTransformedEqualsOperatorSqlCall(SqlCall sqlCall) {
+    List<SqlNode> updatedOperands = new ArrayList<>();
+
+    final SqlTypeNameSpec varcharTypeNameSpec = new SqlBasicTypeNameSpec(SqlTypeName.VARCHAR, ZERO);
+    SqlDataTypeSpec varcharSqlDataTypeSpec = new SqlDataTypeSpec(varcharTypeNameSpec, ZERO);
+
+    for (SqlNode operand : sqlCall.getOperandList()) {
+      SqlNode newOperand = TrinoTryCastFunction.INSTANCE.createCall(POS,
+          new ArrayList<>(Arrays.asList(operand, varcharSqlDataTypeSpec)));
+      updatedOperands.add(newOperand);
+    }
+
+    return sqlCall.getOperator().createCall(POS, updatedOperands);
+  }
+
+  /**
+   * Input SqlCall to this method will access nested fields as is, such as SELECT a.b, a.c.
+   * The below transformations appends an AS operator on nested fields and updated the SqlCall to: SELECT a.b AS b, a.c AS c
+   * @param sqlCall input SqlCall
+   * @return transformed SqlCall
+   */
+  private static SqlCall getTransformedSqlSelectSqlCall(SqlCall sqlCall) {
+    if (((SqlSelect) sqlCall).getSelectList() != null && ((SqlSelect) sqlCall).getSelectList().size() != 0) {
```

Comment:
Does Coral use Apache Commons Collection? If yes, I think it is better to call `Collections.isEmpty()` to check if the list is null or empty().

Reply:
Tried changing to:
`if (CollectionUtils.isNotEmpty(((SqlSelect) sqlCall).getSelectList().getList()))`

it throws NPE since `getSelectList()` might be null. `getSelectList()` returns type `SqlNodeList` which would also need to be checked.